File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    n, m := len(matrix), len(matrix[0])
    start, end := 0, n * m-1

    for start <= end {
        mid := start + (end - start) / 2
        row, col := mid / m, mid % m    // Calculate the row and column indices

        if matrix[row][col] > target {
            end = mid - 1 // Adjust the end index for the left half
        } else if matrix[row][col] < target {
            start = mid + 1 // Adjust the start index for the right half
        } else {
            return true // Found the target value in the matrix
        }
    }

    return false
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length, m = matrix[0].length;
        int low = 0, high = n * m - 1;

        while(low <= high) {
            int mid = low + (high - low) / 2;
            int row = mid / m, col = mid % m;
            if(matrix[row][col] == target) {
                return true;
            }
            if(matrix[row][col] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return false;
    }
}